### ISSUE TITLE
BAD_ENCODING: 2004 error when switching text track stpp -> vtt

### ISSUE THREAD

**Issue by dberber** [shaka-project/shaka-player#2237]
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes

**What version of Shaka Player are you using?**
2.5.5

**Can you reproduce the issue with our latest release version?**
Yes

**Can you reproduce the issue with the latest code from `master`?**
Yes

**Are you using the demo app or your own custom app?**
Our custom app

**If custom app, can you reproduce the issue using our demo app?**
Not tested

**What browser and OS are you using?**
Chrome/Firefox/Edge
MacOS/Win10

**For embedded devices (smart TVs, etc.), what model and firmware version are you using?**

**What are the manifest and license server URIs?**
Internal URIs

**What did you do?**
If you play a live stream which contains text tracks with 2 different mime types where one of them contains binary header in a segment (i.e. stpp with mime: application/mp4) and the second one is a pure text (i.e. vtt with mime: text/vtt) then if you switch from the stpp one to the vtt sometimes (can be very rare) you'll get BAD_ENCODING: 2004 error.

**What did you expect to happen?**
No errors should happen when switching stpp -> vtt.

**What actually happened?**
The 2004 error has happened.

I've analyzed and debugged the shaka's code and here what's happening:
This situation has roots in fetchAndAppend_ method in lib/media/streaming_engine.js and happens when we still have the last stpp segment pending but lib/media/media_source_engine.js has already reinitialized a text parser to use lib/text/vtt_text_parser.js (text track has been switched) and we finally got the last (before switching) stpp segment and passed it down to the vtt text parser then the last-mentioned doesn't expect binary header in the segment data and fails when trying to treat is as a pure utf8 text. So I came up with the fix which successfully resolves this situation the way it drops the last segment of the previous text track if it has different full mime type:
```
// In fetchAndAppend_ method in lib/media/streaming_engine.js:
...
try {
      const results = await Promise.all([initSourceBuffer, fetchSegment]);
      this.destroyer_.ensureNotDestroyed();
      if (this.fatalError_) {
        return;
      }
      // FIX STARTS HERE
      let mediaSourceUpdateIsSafeOperation = true;

      // Check if a mime type has changed while fetching init segment
      // of a text track due to a switch. It's not safe to call appendBuffer
      // in that case (i.e. in case of switching stpp -> vtt) so
      // we just waste the buffer.

      if (mediaState.type === ContentType.TEXT) {
        const currentFullMimeType = shaka.util.MimeUtils.getFullType(
            mediaState.stream.mimeType, mediaState.stream.codecs);
        mediaSourceUpdateIsSafeOperation =
            startingFullMimeType === currentFullMimeType;
      }

      if (mediaSourceUpdateIsSafeOperation) {
        await this.append_(mediaState,
            presentationTime, currentPeriod, stream, reference, results[1]);
      } else {
        // eslint-disable-next-line max-len
        shaka.log.v2(logPrefix, 'cancelling appending the given segment due to switch');
      }
      // FIX ENDS HERE
      this.destroyer_.ensureNotDestroyed();
      if (this.fatalError_) {
        return;
      }
      ...
```

```
// In initSourceBuffer_ method in lib/media/streaming_engine.js:
...
if (!mediaState.needInitSegment) {
      return;
    }
    // FIX STARTS HERE
    const ContentType = shaka.util.ManifestParserUtils.ContentType;

    const startingFullMimeType = shaka.util.MimeUtils.getFullType(
        mediaState.stream.mimeType, mediaState.stream.codecs);
      // FIX ENDS HERE
     ...
      try {
        const initSegment = await fetchInit;
        this.destroyer_.ensureNotDestroyed();
        shaka.log.v1(logPrefix, 'appending init segment');
        // FIX STARTS HERE
        let mediaSourceUpdateIsSafeOperation = true;

        // Check if a mime type has changed while fetching init segment
        // of a text track due to a switch. It's not safe to call appendBuffer
        // in that case (i.e. in case of switching stpp -> vtt) so
        // we just waste the buffer.

        if (mediaState.type === ContentType.TEXT) {
          const currentFullMimeType = shaka.util.MimeUtils.getFullType(
              mediaState.stream.mimeType, mediaState.stream.codecs);
          mediaSourceUpdateIsSafeOperation =
              startingFullMimeType === currentFullMimeType;
        }

        if (mediaSourceUpdateIsSafeOperation) {
          const hasClosedCaptions = mediaState.stream.closedCaptions &&
              mediaState.stream.closedCaptions.size > 0;
          await this.playerInterface_.mediaSourceEngine.appendBuffer(
              mediaState.type, initSegment, null /* startTime */,
              null /* endTime */, hasClosedCaptions);
        } else {
          shaka.log.v2(logPrefix, 'cancelling appending the init segment ' +
              'due to switch');
        }
      // FIX ENDS HERE
      } catch (error) {
        mediaState.needInitSegment = true;
        mediaState.lastInitSegmentReference = null;
        throw error;
      }
      ...
```
I can create a PR if this approach seems correct for the shaka team.

**Comment by TheModMaker** [comment]
That seems like quite a hack to fix this.  I think the issue is that the switch happens while the init segment is being fetched.  A better fix would be to either reinit text as part of the update to avoid this synchronization, or to just check if the stream switched after fetching the segments (this could apply to the media segment append too).

This will probably get fixed as part of period flattening, so we may just wait until #1339 is done.

**Comment by joeyparrish** [comment]
I wonder if we could do something to mirror the flow of audio/video instead.  This doesn't happen with audio/video because there, the decoder only changes with the appending of a new init segment.  With text streams, if we wait to change the text parser (equivalent to an A/V decoder) until we append the new segment, this might be easy to fix.

**Comment by dberber** [comment]
> I wonder if we could do something to mirror the flow of audio/video instead. This doesn't happen with audio/video because there, the decoder only changes with the appending of a new init segment. With text streams, if we wait to change the text parser (equivalent to an A/V decoder) until we append the new segment, this might be easy to fix.

Ok. I think I'll have some time to check if this approach works. I was also wondering why video/audio work but didn't have time/suitable streams to dig deeper. The thing is I've seen two cases when it happens. The first one you're right guys - when it's pending initial segment. But the second one happens when it's pending the next one after initial. That's why I had to fix it in two places.

**Comment by dberber** [comment]
@joeyparrish Changing decoder with the appending of a new init/first (i.e. vtt doesn't have an init segment) segment seems to be working fine, need to fix the unit tests as well. I'll post the fix a bit later, quite busy at work atm.

**Comment by dberber** [comment]
So idea of the fix is the following:

```
// In lib/media/streaming_engine.js
  ...
  switchInternal_(stream, clearBuffer, safeMargin, force) {
    if (mediaState.stream == stream && !force) {
      const streamTag = shaka.media.StreamingEngine.logPrefix_(mediaState);
      shaka.log.debug('switch: Stream ' + streamTag + ' already active');
      return;
    }
    // <-- FIX STARTS
    if (stream.type == ContentType.TEXT) {
      // Mime types are allowed to change for text streams.
      // Schedule reinitializing of a text parser
      // when initial/first segment of the next text track is
      // about to be appended
      mediaState.waitingToReinitializeTextEngine = true;
    }
    // <-- FIX ENDS
    mediaState.stream = stream;
    mediaState.needInitSegment = true;
  }
  ...
  createMediaState_(stream, needPeriodIndex, resumeAt) {
    return /** @type {shaka.media.StreamingEngine.MediaState_} */ ({
      stream: stream,
      type: stream.type,
      lastStream: null,
      lastSegmentReference: null,
      lastInitSegmentReference: null,
      restoreStreamAfterTrickPlay: null,
      needInitSegment: true,
      needPeriodIndex: needPeriodIndex,
      endOfStream: false,
      performingUpdate: false,
      updateTimer: null,
      waitingToClearBuffer: false,
      // <-- FIX STARTS
      waitingToReinitializeTextEngine: false,
      // <-- FIX ENDS
      clearBufferSafeMargin: 0,
      waitingToFlushBuffer: false,
      clearingBuffer: false,
      recovering: false,
      hasError: false,
      resumeAt: resumeAt || 0,
      operation: null,
    });
  }
  ...
  async fetchAndAppend_(
      mediaState, presentationTime, currentPeriodIndex, reference) {
    const ContentType = shaka.util.ManifestParserUtils.ContentType;
    const StreamingEngine = shaka.media.StreamingEngine;
    const logPrefix = StreamingEngine.logPrefix_(mediaState);
    const currentPeriod = this.manifest_.periods[currentPeriodIndex];
      // <-- FIX STARTS
    // Get waitingToReinitializeTextEngine right here
    // before all asynchronous calls to ensure
    // it hasn't been changed by switchInternal_
    const waitingToReinitializeTextEngine =
        mediaState.waitingToReinitializeTextEngine;
      // <-- FIX ENDS
    ...
    try {
      const results = await Promise.all([initSourceBuffer, fetchSegment]);
      this.destroyer_.ensureNotDestroyed();
      if (this.fatalError_) {
        return;
      }
      // <-- FIX STARTS
      if (waitingToReinitializeTextEngine) {
        const fullMimeType = shaka.util.MimeUtils.getFullType(
            mediaState.stream.mimeType, mediaState.stream.codecs);
        this.playerInterface_.mediaSourceEngine.reinitText(fullMimeType);

        mediaState.waitingToReinitializeTextEngine = false;
      }
      // <-- FIX ENDS
      await this.append_(mediaState,
          presentationTime, currentPeriod, stream, reference, results[1]);
      this.destroyer_.ensureNotDestroyed();
      if (this.fatalError_) {
        return;
      }
      ...
  }
  ...
  async initSourceBuffer_(
      mediaState, currentPeriodIndex, appendWindowStart, appendWindowEnd,
      reference) {
    // TODO: Remove needInitSegment.  Currently, this both signals the need for
    // a different init segment (switches, period transitions) and protects
    // against unnecessary calls to setStreamProperties.  If we can solve calls
    // to setStreamProperties another way, then we could finally drop
    // needInitSegment.
    if (!mediaState.needInitSegment) {
      return;
    }
    // <-- FIX STARTS
    const waitingToReinitializeTextEngine =
        mediaState.waitingToReinitializeTextEngine;
    // <-- FIX ENDS
    ...
    const append = async () => {
      try {
        const initSegment = await fetchInit;
        this.destroyer_.ensureNotDestroyed();
        shaka.log.v1(logPrefix, 'appending init segment');
        const hasClosedCaptions = mediaState.stream.closedCaptions &&
            mediaState.stream.closedCaptions.size > 0;
        // <-- FIX STARTS
        if (waitingToReinitializeTextEngine) {
          const fullMimeType = shaka.util.MimeUtils.getFullType(
              mediaState.stream.mimeType, mediaState.stream.codecs);
          this.playerInterface_.mediaSourceEngine.reinitText(fullMimeType);

          mediaState.waitingToReinitializeTextEngine = false;
        }
        // <-- FIX ENDS
      ...
    ...
  }
```

**Comment by dberber** [comment]
@joeyparrish does't make sense to do a PR based on this approach?

**Comment by TheModMaker** [comment]
I think a better fix would just to call `reinitText` always inside `initSourceBuffer_`.  Since that is called when we initialize the stream, we just need to initialize the text stream there too.  You could store the current `mediaState.stream` value before fetching the init segment and make sure we didn't switch while fetching.

Also, interesting feature, you can post a diff on GitHub and it will display it colored.   You can just use something like `git diff` and just post that. Just use <code>```diff</code>:

```diff
diff --git a/lib/media/streaming_engine.js b/lib/media/streaming_engine.js
index 71f6d474..7048955e 100644
--- a/lib/media/streaming_engine.js
+++ b/lib/media/streaming_engine.js
@@ -553,15 +553,6 @@ shaka.media.StreamingEngine = class {
       return;
     }
 
-    if (stream.type == ContentType.TEXT) {
-      // Mime types are allowed to change for text streams.
-      // Reinitialize the text parser, but only if we are going to fetch the
-      // init segment again.
-      const fullMimeType = shaka.util.MimeUtils.getFullType(
-          stream.mimeType, stream.codecs);
-      this.playerInterface_.mediaSourceEngine.reinitText(fullMimeType);
-    }
-
     mediaState.stream = stream;
     mediaState.needInitSegment = true;
 
@@ -1623,6 +1614,12 @@ shaka.media.StreamingEngine = class {
       return;
     }
 
+    if (mediaState.type == shaka.util.ManifestParserUtils.ContentType.TEXT) {
+      const fullMimeType = shaka.util.MimeUtils.getFullType(
+          mediaState.stream.mimeType, mediaState.stream.codecs);
+      this.playerInterface_.mediaSourceEngine.reinitText(fullMimeType);
+    }
+
     const logPrefix = shaka.media.StreamingEngine.logPrefix_(mediaState);
     const currentPeriod = this.manifest_.periods[currentPeriodIndex];
 
```

**Comment by joeyparrish** [comment]
@dberber, does @TheModMaker's suggestion work for you?

**Comment by dberber** [comment]
@joeyparrish @TheModMaker I'll check on a real stream on Monday. If I'm not wrong there is an edge case where it doesn't go futher than the first if in initSourceBuffer_ method because a vtt track doesn't have an init segment. That's why I've amended the both methods. @TheModMaker thanks for the hint on the feature.

**Comment by michellezhuogg** [comment]
Hello @dberber ,

Are you still working on fix this issue? Or, would you like to provide a manifest for us to reproduce the problem and work on a fix? Thanks!

**Comment by dberber** [comment]
Hello @michellezhuogg,
I've done the fix and it's on a probation at the moment. Everything has been fine under heavy usage until today suddenly I've got a related error (2007 INVALID_MP4_TTML). So there is a lot to investigate again and I think it's related to some changes engineers have made to the streams. Unfortunately I can't provide the manifests as they're private, DRM'ed and I've signed a NDA. But we have a release soon and I have every intention of fixing this up as soon as possible and of course I'll share my findings and fixes with the community as well.

**Comment by joeyparrish** [comment]
Sounds good.  If you have a fix to share with us, please feel free to send a PR.  In the mean time, I will move this issue to the backlog.

**Comment by joeyparrish** [comment]
Or, I would move this to the backlog, if we didn't have an open ticket with GitHub support right now about how we can't modify our backlog milestone.  :-(

**Comment by dberber** [comment]
Hello @joeyparrish.
So after a couple of months of heavy testing and addtional fixing here is the fix with explanations.

```diff
diff --git a/lib/media/streaming_engine.js b/lib/media/streaming_engine.js
index b92457c4..13fcc79a 100644
--- a/lib/media/streaming_engine.js
+++ b/lib/media/streaming_engine.js
@@ -543,11 +543,11 @@ shaka.media.StreamingEngine = class {
 
     if (stream.type == ContentType.TEXT) {
       // Mime types are allowed to change for text streams.
-      // Reinitialize the text parser, but only if we are going to fetch the
-      // init segment again.
-      const fullMimeType = shaka.util.MimeUtils.getFullType(
-          stream.mimeType, stream.codecs);
-      this.playerInterface_.mediaSourceEngine.reinitText(fullMimeType);
+      // Set needTextParserReinitialisation flag in order to
+      // postpone reinitialization of the text parser until the moment
+      // we are about to fetch an init segment again. It prevents:
+      // https://github.com/google/shaka-player/issues/2237
+      mediaState.needTextParserReinitialisation = true;
     }
 
     mediaState.stream = stream;
@@ -893,6 +893,7 @@ shaka.media.StreamingEngine = class {
       lastInitSegmentReference: null,
       restoreStreamAfterTrickPlay: null,
       needInitSegment: true,
+      needTextParserReinitialisation: false,
       needPeriodIndex: needPeriodIndex,
       endOfStream: false,
       performingUpdate: false,
@@ -1574,6 +1575,9 @@ shaka.media.StreamingEngine = class {
       return;
     }
 
+    const needTextParserReinitialisation =
+      mediaState.needTextParserReinitialisation;
+
     // If we need an init segment, then the Stream switched, so we've either
     // changed bitrates, Periods, or both. If we've changed Periods then we must
     // set a new timestamp offset and append window end. Note that by setting
@@ -1589,8 +1593,18 @@ shaka.media.StreamingEngine = class {
             mediaState.type, timestampOffset, appendWindowStart,
             appendWindowEnd);
 
+    if (needTextParserReinitialisation) {
+      const fullMimeType = shaka.util.MimeUtils.getFullType(
+          mediaState.stream.mimeType, mediaState.stream.codecs);
+      shaka.log.v1(logPrefix, 'reinitializing a text parser');
+      this.playerInterface_.mediaSourceEngine.reinitText(fullMimeType);
+      mediaState.needTextParserReinitialisation = false;
+    }
+
     if (reference.initSegmentReference == mediaState.lastInitSegmentReference) {
       // The SourceBuffer already has the correct init segment appended.
+      shaka.log.v1(logPrefix,
+          'the SourceBuffer already has the correct init segment appended');
       await setStreamProperties;
       return;
     }
@@ -1599,6 +1613,7 @@ shaka.media.StreamingEngine = class {
 
     if (!reference.initSegmentReference) {
       // The Stream is self initializing.
+      shaka.log.v1(logPrefix, 'the Stream is self initializing');
       await setStreamProperties;
       return;
     }
@@ -2347,6 +2362,7 @@ shaka.media.StreamingEngine.PlayerInterface;
  *   lastInitSegmentReference: shaka.media.InitSegmentReference,
  *   restoreStreamAfterTrickPlay: ?shaka.extern.Stream,
  *   needInitSegment: boolean,
+ *   needTextParserReinitialisation,
  *   needPeriodIndex: number,
  *   endOfStream: boolean,
  *   performingUpdate: boolean,
diff --git a/test/media/streaming_engine_unit.js b/test/media/streaming_engine_unit.js
index 5f62418b..aac009a1 100644
--- a/test/media/streaming_engine_unit.js
+++ b/test/media/streaming_engine_unit.js
@@ -503,8 +503,6 @@ describe('StreamingEngine', () => {
         verifyNetworkingEngineRequestCalls(1);
 
         onCanSwitch.and.callFake(() => {
-          expect(mediaSourceEngine.reinitText).toHaveBeenCalled();
-          mediaSourceEngine.reinitText.calls.reset();
           onCanSwitch.and.throwError(new Error());
         });
 
@@ -535,6 +533,9 @@ describe('StreamingEngine', () => {
     streamingEngine.start();
 
     await runTest();
+    // Should be called only once before applying
+    // an init segment of the second period
+    expect(mediaSourceEngine.reinitText).toHaveBeenCalledTimes(1);
     expect(mediaSourceEngine.endOfStream).toHaveBeenCalled();
 
     // Verify buffers.
@@ -840,7 +841,6 @@ describe('StreamingEngine', () => {
       onCanSwitch.and.callFake(() => {
         expect(streamingEngine.getBufferingText()).toBe(textStream2);
 
-        mediaSourceEngine.reinitText.calls.reset();
         streamingEngine.switchTextStream(textStream2);
       });
     });
@@ -849,7 +849,10 @@ describe('StreamingEngine', () => {
     streamingEngine.start();
     await runTest();
 
-    expect(mediaSourceEngine.reinitText).not.toHaveBeenCalled();
+    // Should be called only once after it's switched from the first period
+    // but switching to the already playing text track when
+    // in the second period should be ignored
+    expect(mediaSourceEngine.reinitText).toHaveBeenCalledTimes(1);
   });
 
   it('plays when 2nd Period doesn\'t have text streams', async () => {
```

So the main idea of this fix is to postpone reinitialization of the text parser until the moment we are about to fetch an init segment again. This approach and thorough log analysis has proved that the text parser is reinitialized at the right moment when we're ready to get and consume the init segment of the text track we're switched to. The only thing happend recently and needed amendment is - a very rare edge case when we're starting with MP4 stpp text track and almost instantly switching to a vtt one: since the text track buffer is still empty at the moment of switching so this condition: if (reference.initSegmentReference == mediaState.lastInitSegmentReference) is true because both properties are null and that led to the INVALID_MP4_TTML 2007 error because it tried to feed mp4 ttml parser with vtt data (the other way around situation than we tried to fix in the beggining) so I had to move the reinitialization piece of the code up one line before this condition.

**Comment by mimse** [comment]
I made a comment on the PR https://github.com/google/shaka-player/pull/2425#issuecomment-789529233

**Comment by joeyparrish** [comment]
PR #2425 looks like a good fix for v2.5.x, but v3.0.x and v3.1.x look totally different and will need to have the PR ported to those branches.

**Comment by avelad** [comment]
Tested in the main branch and now it works as expected
